### ISSUE TITLE
Wrong behavior when `ENUM` column meet Aggregation

### ISSUE THREAD

**Issue by winoros** [pingcap/tidb#13027]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a enum('a', 'b'));
insert into t values('a'), ('b');
select cast(a as signed) from (select a from t group by a) t;
```

2. What did you expect to see?

Two row. 1 and 2.

3. What did you see instead?

Two row. Both are 0, with warning `Data Truncated`

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

current master 02b7634954aa576e1c176db6744b26d641b0a6ce

**Comment by winoros** [comment]
`max(a)` and `min(a)` should also return the correct result without warning.

**Comment by pingyu** [comment]
Is it caused by [`(*baseFuncDesc).typeInfer4MaxMin`](https://github.com/pingcap/tidb/blob/b743208fd72dec3c7fc3532d1b627329537a5bbf/expression/aggregation/base_func.go#L192) ?
```go
if a.RetTp.Tp == mysql.TypeEnum || a.RetTp.Tp == mysql.TypeSet {
	a.RetTp = &types.FieldType{Tp: mysql.TypeString, Flen: mysql.MaxFieldCharLength}
}
```

So we should add a `typeInfer4FirstRow`, which will infer to a same type as argument. Shouldn't we ?

PS. I found another issue when `ENUM` met aggregation: #13961

**Comment by winoros** [comment]
I think the one you found shares the same cause of this issue. You can have a try to do a `cast as signed` on the result after MAX/MIN.
If these two are the same cause, no need to add a new `typeInfer4FirstRow`. Change the original `typeInfer4MaxMin` is enough. (Also we need to change the execution logic).

**Comment by pingyu** [comment]
@winoros 
Thanks for your advise. I had tried to fix it (#14035), would you please take a look ?
